### PR TITLE
fix(terraform): Reduce constraints changes in lockfiles

### LINKED ISSUE #25193: Terraform: reduce constraints changes
### Describe the proposed change(s).

When currentValue===newValue then don't change the `constraints` in the lock file at all.

Otherwise, if the current constraints contains `currentValue` then replace it with `newValue` (leaving rest as-is).

Otherwise, if the current constraints contains `currentVersion` then replace it with `newVersion` (leaving rest as-is).

Otherwise, fall back to old functionality (rewriting complete constraints value).

### PR DESCRIPTION
**Pull request by zharinov**
## Context

- Closes: #25193

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/terraform/lockfile/index.spec.ts [hunk 1]
```diff
@@ -910,4 +910,183 @@ describe('modules/manager/terraform/lockfile/index', () => {
...
+      packageFileName: 'main.tf',
+      updatedDeps: [
+        {
+          depName: 'aws',
+          depType: 'provider',
+          packageName: 'hashicorp/aws',
+          registryUrls: ['https://registry.example.com'],
+          newVersion: '3.37.0',
+          currentValue: '~> 3.0.0',
+          newValue: '~> 3.37.0',
```
Comment (secustor) [inline lib/modules/manager/terraform/lockfile/index.spec.ts:1003]:
```suggestion
          newValue: '~> 3.37',
```
Just to test the number full version is templated.

File: lib/modules/manager/terraform/lockfile/index.ts [hunk 2]
```diff
@@ -99,19 +99,66 @@ export async function updateArtifacts({
...
           : TerraformProviderDatasource.defaultRegistryUrls[0];
-        const newConstraint = isPinnedVersion(newValue) ? newVersion : newValue;
         const updateLock = locks.find(
           (value) => value.packageName === packageName
         );
         // istanbul ignore if: needs test
         if (!updateLock) {
           continue;
         }
+        let newConstraint: string | undefined = updateLock.constraints;
```
Comment (secustor) [inline lib/modules/manager/terraform/lockfile/index.ts:121]: Please move this in a separate function for better readability.

**Review by viceice — COMMENTED** [review]
looks good, but I think this needs real repo tests

**Comment by zharinov** [comment]
Examples:
- https://github.com/renovate-testing/test-25430-terraform/pulls
- https://github.com/renovate-testing/test-25430-terraform-2/pulls

**Review by rarkins — APPROVED** [review]

**Pull request merged**